doc/services/rtio/index.rst

```diff
@@ -0,0 +1,417 @@
+perhaps the read could be done using DMA as its large enough make sense. That
+requires an understanding of how to setup the device's particular DMA to do so.
+
+The above sequence of operations is embodied in RTIO as chain of
+submission queue entries (sqe). Chaining is done by setting a bitflag in
+an sqe to signify the next sqe must wait on the current one.
+
+Because the chip select and clocking is common to a particular SPI controller
+and device on the bus it is embodied in what RTIO calls an iodev.
+
+Multiple operations against the same iodev are done in the order provided as
+soon as possible. If two operation chains have varying points using the same
+device its possible one chain will have to wait for another to complete.
+
+Completion Queue
+****************
+
+In order to know when a sqe has completed there is a completion
+queue (cq) with completion queue events (cqe). A sqe once completed results in
+a cqe being pushed into the cq. The ordering of cqe may not be the same order of
+sqe. A chain of sqe will however ensure ordering and failure cascading.
+
+Other potential schemes are possible but a completion queue is a well trod
+idea with io_uring and other similar operating system APIs.
+
+Executor and IODev
+******************
+
+Turning submission queue entries (sqe) into completion events (cqe) is the job
+of objects implementing the executor and iodev APIs. These APIs enable
+coordination between themselves to enable things like DMA transfers.
+
+The end result of these APIs should be a method to resolve the request by
+deciding some of the following questions with heuristic/constraint
+based decision making.
+
+* Polling, Interrupt, or DMA transfer?
+* If DMA, are the requirements met (peripheral supported by DMAC, etc).
+* Block/Non-Blocking
+
+Outstanding Questions
+*********************
+
+RTIO is not a complete API and solution, and is currently evolving to best
+fit the nature of an RTOS. The general ideas behind a pair of queues to
+describe requests and completions seems sound and has been proven out in
+other contexts. Questions remain though.
+
+Timeouts and Deadlines
+======================
+
+Timeouts and deadlines are key to being Real-Time. Real-Time in Zephyr means
+being able to do things when an application wants them done. That could mean
+different things from a deadline with best effort attempts or a timeout and
+failure.
+
+These features would surely be useful in many cases, but would likely add some
+significant complexities. It's something to decide upon, and even if enabled
+would likely be an compile time optional feature leading to complex testing.
+
+Cancellation
+============
+
+Canceling an already queued operation could be possible with a small
+API addition to perhaps take both the RTIO context and a pointer to the
+submission queue entry. However, cancellation as an API induces many potential
+complexities that might not be appropriate. It's something to be decided upon.
```

Comment:
Canceling may make sense at least for times or failures and timeouts. This would free the resource of wasting time in case another SQE is waiting to run on that device

Reply:
Yes, though has some potential implications. Right now the sqe is setup in a pair of acquire/produce calls which wrap the safe area in which to do mutation of a sqe from the application. By introducing cancel, there needs to be a way to continue mutating the sqe in some safe form, likely with an atomic bit flag that gets reset on acquire again (so its a best effort cancel, not a blocking locking cancel). There's also some memory usage additions from cancel then for each SQE. An ifdef would seem appropriate as not everyone is likely to care about it.

Reply:
I think best effort is better than possibly hanging up a device for no reason. I am always good with ifdef reducing footprint on unneeded features.